### PR TITLE
[OPPRO-311] Generate micro benchmark files and enable in gha.

### PR DESCRIPTION
**Pull request by marin-ma**
Developers who want to generate input json file and data files for micro benchmak may refer to test("Generate example substrait plan") in [NativeBenchmarkPlanGenerator.scala](https://github.com/oap-project/gluten/pull/556/files#diff-f8eb8bf4c7ae74a2ffb260dd5d9dad9353c592a91344b395f667e928145300fb
)
This workload generates example input json plan and data files in gluten/velox-backend/generated-native-benchmark. To run micro benchmark with examples:
```
cd gluten/cpp/velox/benchmarks
./generic_benchmark --iterations 1 --threads 1 /path/to/gluten/velox-backend/generated-native-benchmark/example.json /path/to/gluten/velox-backend/generated-native-benchmark/example_lineitem/part-xxxx.parquet /path/to/gluten/velox-backend/generated-native-benchmark/example_orders/part-xxxx.parquet
```

Also supports printing substrait json via df.explain(true). df.explain() or df.explain(false) will not print the json.
Here's the printed plan for TPCH Q4:

<details>
<summary>q4.exlain(true)</summary>

```
== Parsed Logical Plan ==
'Sort ['o_orderpriority ASC NULLS FIRST], true
+- 'Aggregate ['o_orderpriority], ['o_orderpriority, 'count(1) AS order_count#122]
   +- 'Filter ((('o_orderdate >= 1993-07-01) AND ('o_orderdate < 1993-10-01)) AND exists#123 [])
      :  +- 'Project [*]
      :     +- 'Filter (('l_orderkey = 'o_orderkey) AND ('l_commitdate < 'l_receiptdate))
      :        +- 'UnresolvedRelation [lineitem], [], false
      +- 'UnresolvedRelation [orders], [], false

== Analyzed Logical Plan ==
o_orderpriority: string, order_count: bigint
Sort [o_orderpriority#61 ASC NULLS FIRST], true
+- Aggregate [o_orderpriority#61], [o_orderpriority#61, count(1) AS order_count#122L]
   +- Filter (((o_orderdate#60 >= 1993-07-01) AND (o_orderdate#60 < 1993-10-01)) AND exists#123 [o_orderkey#56L])
      :  +- Project [l_orderkey#16L, l_partkey#17L, l_suppkey#18L, l_linenumber#19, l_quantity#20, l_extendedprice#21, l_discount#22, l_tax#23, l_returnflag#24, l_linestatus#25, l_shipdate#26, l_commitdate#27, l_receiptdate#28, l_shipinstruct#29, l_shipmode#30, l_comment#31]
      :     +- Filter ((l_orderkey#16L = outer(o_orderkey#56L)) AND (l_commitdate#27 < l_receiptdate#28))
      :        +- SubqueryAlias lineitem
      :           +- View (`lineitem`, [l_orderkey#16L,l_partkey#17L,l_suppkey#18L,l_linenumber#19,l_quantity#20,l_extendedprice#21,l_discount#22,l_tax#23,l_returnflag#24,l_linestatus#25,l_shipdate#26,l_commitdate#27,l_receiptdate#28,l_shipinstruct#29,l_shipmode#30,l_comment#31])
      :              +- RelationV2[l_orderkey#16L, l_partkey#17L, l_suppkey#18L, l_linenumber#19, l_quantity#20, l_extendedprice#21, l_discount#22, l_tax#23, l_returnflag#24, l_linestatus#25, l_shipdate#26, l_commitdate#27, l_receiptdate#28, l_shipinstruct#29, l_shipmode#30, l_comment#31] parquet hdfs://10.112.242.189:9000/tpch_sf2t_parquet_192p/lineitem
      +- SubqueryAlias orders
         +- View (`orders`, [o_orderkey#56L,o_custkey#57L,o_orderstatus#58,o_totalprice#59,o_orderdate#60,o_orderpriority#61,o_clerk#62,o_shippriority#63,o_comment#64])
            +- RelationV2[o_orderkey#56L, o_custkey#57L, o_orderstatus#58, o_totalprice#59, o_orderdate#60, o_orderpriority#61, o_clerk#62, o_shippriority#63, o_comment#64] parquet hdfs://10.112.242.189:9000/tpch_sf2t_parquet_192p/orders

== Optimized Logical Plan ==
Sort [o_orderpriority#61 ASC NULLS FIRST], true
+- Aggregate [o_orderpriority#61], [o_orderpriority#61, count(1) AS order_count#122L]
   +- Project [o_orderpriority#61]
      +- Join LeftSemi, (l_orderkey#16L = o_orderkey#56L)
         :- Project [o_orderkey#56L, o_orderpriority#61]
         :  +- Filter ((isnotnull(o_orderdate#60) AND (o_orderdate#60 >= 1993-07-01)) AND (o_orderdate#60 < 1993-10-01))
         :     +- RelationV2[o_orderkey#56L, o_orderdate#60, o_orderpriority#61] parquet hdfs://10.112.242.189:9000/tpch_sf2t_parquet_192p/orders
         +- Project [l_orderkey#16L]
            +- Filter (((l_commitdate#27 < l_receiptdate#28) AND isnotnull(l_commitdate#27)) AND isnotnull(l_receiptdate#28))
               +- RelationV2[l_orderkey#16L, l_partkey#17L, l_suppkey#18L, l_linenumber#19, l_quantity#20, l_extendedprice#21, l_discount#22, l_tax#23, l_returnflag#24, l_linestatus#25, l_shipdate#26, l_commitdate#27, l_receiptdate#28, l_shipinstruct#29, l_shipmode#30, l_comment#31] parquet hdfs://10.112.242.189:9000/tpch_sf2t_parquet_192p/lineitem

== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=true
+- == Final Plan ==
   VeloxNativeColumnarToRowExec
   +- *(5) SortExecTransformer [o_orderpriority#61 ASC NULLS FIRST], true, 0
      +- CoalesceBatches
         +- ColumnarAQEShuffleRead [CoalescedPartitionSpec(0,384,Some(1255))]
            +- ShuffleQueryStage 3
               +- ColumnarExchangeAdaptor rangepartitioning(o_orderpriority#61 ASC NULLS FIRST, 384), ENSURE_REQUIREMENTS, false, [plan_id=268], [id=#268], [OUTPUT] List(o_orderpriority:StringType, order_count:LongType), [OUTPUT] List(o_orderpriority:StringType, order_count:LongType)
                  +- *(4) HashAggregateTransformer(keys=[o_orderpriority#61], functions=[count(1)], output=[o_orderpriority#61, order_count#122L])
                     +- CoalesceBatches
                        +- ColumnarAQEShuffleRead [CoalescedPartitionSpec(0,384,Some(481920))]
                           +- ShuffleQueryStage 2
                              +- ColumnarExchangeAdaptor hashpartitioning(o_orderpriority#61, 384), ENSURE_REQUIREMENTS, true, [plan_id=227], [id=#227], [OUTPUT] List(o_orderpriority:StringType, count:LongType), [OUTPUT] List(o_orderpriority:StringType, count:LongType)
                                 +- *(3) ProjectExecTransformer [hash(o_orderpriority#61, 42) AS hash_partition_key#172, o_orderpriority#61, count#169L]
                                    +- *(3) HashAggregateTransformer(keys=[o_orderpriority#61], functions=[partial_count(1)], output=[o_orderpriority#61, count#169L])
                                       +- *(3) ProjectExecTransformer [o_orderpriority#61]
                                          +- *(3) VeloxShuffledHashJoinExecTransformer [o_orderkey#56L], [l_orderkey#16L], LeftSemi, BuildLeft
                                             :- ShuffleQueryStage 0
                                             :  +- ColumnarExchangeAdaptor hashpartitioning(o_orderkey#56L, 384), ENSURE_REQUIREMENTS, true, [plan_id=110], [id=#110], [OUTPUT] List(o_orderkey:LongType, o_orderpriority:StringType), [OUTPUT] List(o_orderkey:LongType, o_orderpriority:StringType)
                                             :     +- *(1) ProjectExecTransformer [hash(o_orderkey#56L, 42) AS hash_partition_key#170, o_orderkey#56L, o_orderpriority#61]
                                             :        +- *(1) VeloxFilterExecTransformer ((isnotnull(o_orderdate#60) AND (o_orderdate#60 >= 1993-07-01)) AND (o_orderdate#60 < 1993-10-01))
                                             :           +- *(1) BatchScanExecTransformer[o_orderkey#56L, o_orderdate#60, o_orderpriority#61] ParquetScan DataFilters: [isnotnull(o_orderdate#60), (o_orderdate#60 >= 1993-07-01), (o_orderdate#60 < 1993-10-01)], Format: parquet, Location: InMemoryFileIndex(1 paths)[hdfs://10.112.242.189:9000/tpch_sf2t_parquet_192p/orders], PartitionFilters: [], PushedFilters: [IsNotNull(o_orderdate), GreaterThanOrEqual(o_orderdate,1993-07-01), LessThan(o_orderdate,1993-10..., ReadSchema: struct<o_orderkey:bigint,o_orderdate:string,o_orderpriority:string>, PushedFilters: [IsNotNull(o_orderdate), GreaterThanOrEqual(o_orderdate,1993-07-01), LessThan(o_orderdate,1993-10-01)] RuntimeFilters: []
*(1) Substrait plan: {
  "extensions": [{
    "extensionFunction": {
      "functionAnchor": 1,
      "name": "gte:str_str"
    }
  }, {
    "extensionFunction": {
      "name": "is_not_null:str"
    }
  }, {
    "extensionFunction": {
      "functionAnchor": 4,
      "name": "murmur3hash:opt_i64"
    }
  }, {
    "extensionFunction": {
      "functionAnchor": 3,
      "name": "lt:str_str"
    }
  }, {
    "extensionFunction": {
      "functionAnchor": 2,
      "name": "and:bool_bool"
    }
  }],
  "relations": [{
    "root": {
      "input": {
        "project": {
          "common": {
            "direct": {
            }
          },
          "input": {
            "read": {
              "common": {
                "direct": {
                }
              },
              "baseSchema": {
                "names": ["o_orderkey", "o_orderdate", "o_orderpriority"],
                "struct": {
                  "types": [{
                    "i64": {
                      "nullability": "NULLABILITY_NULLABLE"
                    }
                  }, {
                    "string": {
                      "nullability": "NULLABILITY_NULLABLE"
                    }
                  }, {
                    "string": {
                      "nullability": "NULLABILITY_NULLABLE"
                    }
                  }]
                },
                "partitionColumns": {
                  "columnType": ["NORMAL_COL", "NORMAL_COL", "NORMAL_COL"]
                }
              },
              "filter": {
                "scalarFunction": {
                  "functionReference": 2,
                  "outputType": {
                    "bool": {
                      "nullability": "NULLABILITY_NULLABLE"
                    }
                  },
                  "arguments": [{
                    "value": {
                      "scalarFunction": {
                        "functionReference": 2,
                        "outputType": {
                          "bool": {
                            "nullability": "NULLABILITY_NULLABLE"
                          }
                        },
                        "arguments": [{
                          "value": {
                            "scalarFunction": {
                              "outputType": {
                                "bool": {
                                  "nullability": "NULLABILITY_REQUIRED"
                                }
                              },
                              "arguments": [{
                                "value": {
                                  "selection": {
                                    "directReference": {
                                      "structField": {
                                        "field": 1
                                      }
                                    }
                                  }
                                }
                              }]
                            }
                          }
                        }, {
                          "value": {
                            "scalarFunction": {
                              "functionReference": 1,
                              "outputType": {
                                "bool": {
                                  "nullability": "NULLABILITY_NULLABLE"
                                }
                              },
                              "arguments": [{
                                "value": {
                                  "selection": {
                                    "directReference": {
                                      "structField": {
                                        "field": 1
                                      }
                                    }
                                  }
                                }
                              }, {
                                "value": {
                                  "literal": {
                                    "string": "1993-07-01"
                                  }
                                }
                              }]
                            }
                          }
                        }]
                      }
                    }
                  }, {
                    "value": {
                      "scalarFunction": {
                        "functionReference": 3,
                        "outputType": {
                          "bool": {
                            "nullability": "NULLABILITY_NULLABLE"
                          }
                        },
                        "arguments": [{
                          "value": {
                            "selection": {
                              "directReference": {
                                "structField": {
                                  "field": 1
                                }
                              }
                            }
                          }
                        }, {
                          "value": {
                            "literal": {
                              "string": "1993-10-01"
                            }
                          }
                        }]
                      }
                    }
                  }]
                }
              }
            }
          },
          "expressions": [{
            "scalarFunction": {
              "functionReference": 4,
              "outputType": {
                "i32": {
                  "nullability": "NULLABILITY_REQUIRED"
                }
              },
              "arguments": [{
                "value": {
                  "selection": {
                    "directReference": {
                      "structField": {
                      }
                    }
                  }
                }
              }]
            }
          }, {
            "selection": {
              "directReference": {
                "structField": {
                }
              }
            }
          }, {
            "selection": {
              "directReference": {
                "structField": {
                  "field": 2
                }
              }
            }
          }]
        }
      },
      "names": ["hash_partition_key#170", "o_orderkey#56", "o_orderpriority#61"]
    }
  }]
}
                                             +- ShuffleQueryStage 1
                                                +- ColumnarExchangeAdaptor hashpartitioning(l_orderkey#16L, 384), ENSURE_REQUIREMENTS, true, [plan_id=134], [id=#134], [OUTPUT] List(l_orderkey:LongType), [OUTPUT] List(l_orderkey:LongType)
                                                   +- *(2) ProjectExecTransformer [hash(l_orderkey#16L, 42) AS hash_partition_key#171, l_orderkey#16L]
                                                      +- *(2) VeloxFilterExecTransformer (((l_commitdate#27 < l_receiptdate#28) AND isnotnull(l_commitdate#27)) AND isnotnull(l_receiptdate#28))
                                                         +- *(2) BatchScanExecTransformer[l_orderkey#16L, l_partkey#17L, l_suppkey#18L, l_linenumber#19, l_quantity#20, l_extendedprice#21, l_discount#22, l_tax#23, l_returnflag#24, l_linestatus#25, l_shipdate#26, l_commitdate#27, l_receiptdate#28, l_shipinstruct#29, l_shipmode#30, l_comment#31] ParquetScan DataFilters: [(l_commitdate#27 < l_receiptdate#28), isnotnull(l_commitdate#27), isnotnull(l_receiptdate#28)], Format: parquet, Location: InMemoryFileIndex(1 paths)[hdfs://10.112.242.189:9000/tpch_sf2t_parquet_192p/lineitem], PartitionFilters: [], PushedFilters: [IsNotNull(l_commitdate), IsNotNull(l_receiptdate)], ReadSchema: struct<l_orderkey:bigint,l_partkey:bigint,l_suppkey:bigint,l_linenumber:int,l_quantity:double,l_e..., PushedFilters: [IsNotNull(l_commitdate), IsNotNull(l_receiptdate)] RuntimeFilters: []
*(2) Substrait plan: {
  "extensions": [{
    "extensionFunction": {
      "functionAnchor": 1,
      "name": "is_not_null:str"
    }
  }, {
    "extensionFunction": {
      "functionAnchor": 3,
      "name": "murmur3hash:opt_i64"
    }
  }, {
    "extensionFunction": {
      "name": "lt:str_str"
    }
  }, {
    "extensionFunction": {
      "functionAnchor": 2,
      "name": "and:bool_bool"
    }
  }],
  "relations": [{
    "root": {
      "input": {
        "project": {
          "common": {
            "direct": {
            }
          },
          "input": {
            "read": {
              "common": {
                "direct": {
                }
              },
              "baseSchema": {
                "names": ["l_orderkey", "l_partkey", "l_suppkey", "l_linenumber", "l_quantity", "l_extendedprice", "l_discount", "l_tax", "l_returnflag", "l_linestatus", "l_shipdate", "l_commitdate", "l_receiptdate", "l_shipinstruct", "l_shipmode", "l_comment"],
                "struct": {
                  "types": [{
                    "i64": {
                      "nullability": "NULLABILITY_NULLABLE"
                    }
                  }, {
                    "i64": {
                      "nullability": "NULLABILITY_NULLABLE"
                    }
                  }, {
                    "i64": {
                      "nullability": "NULLABILITY_NULLABLE"
                    }
                  }, {
                    "i32": {
                      "nullability": "NULLABILITY_NULLABLE"
                    }
                  }, {
                    "fp64": {
                      "nullability": "NULLABILITY_NULLABLE"
                    }
                  }, {
                    "fp64": {
                      "nullability": "NULLABILITY_NULLABLE"
                    }
                  }, {
                    "fp64": {
                      "nullability": "NULLABILITY_NULLABLE"
                    }
                  }, {
                    "fp64": {
                      "nullability": "NULLABILITY_NULLABLE"
                    }
                  }, {
                    "string": {
                      "nullability": "NULLABILITY_NULLABLE"
                    }
                  }, {
                    "string": {
                      "nullability": "NULLABILITY_NULLABLE"
                    }
                  }, {
                    "string": {
                      "nullability": "NULLABILITY_NULLABLE"
                    }
                  }, {
                    "string": {
                      "nullability": "NULLABILITY_NULLABLE"
                    }
                  }, {
                    "string": {
                      "nullability": "NULLABILITY_NULLABLE"
                    }
                  }, {
                    "string": {
                      "nullability": "NULLABILITY_NULLABLE"
                    }
                  }, {
                    "string": {
                      "nullability": "NULLABILITY_NULLABLE"
                    }
                  }, {
                    "string": {
                      "nullability": "NULLABILITY_NULLABLE"
                    }
                  }]
                },
                "partitionColumns": {
                  "columnType": ["NORMAL_COL", "NORMAL_COL", "NORMAL_COL", "NORMAL_COL", "NORMAL_COL", "NORMAL_COL", "NORMAL_COL", "NORMAL_COL", "NORMAL_COL", "NORMAL_COL", "NORMAL_COL", "NORMAL_COL", "NORMAL_COL", "NORMAL_COL", "NORMAL_COL", "NORMAL_COL"]
                }
              },
              "filter": {
                "scalarFunction": {
                  "functionReference": 2,
                  "outputType": {
                    "bool": {
                      "nullability": "NULLABILITY_NULLABLE"
                    }
                  },
                  "arguments": [{
                    "value": {
                      "scalarFunction": {
                        "functionReference": 2,
                        "outputType": {
                          "bool": {
                            "nullability": "NULLABILITY_NULLABLE"
                          }
                        },
                        "arguments": [{
                          "value": {
                            "scalarFunction": {
                              "outputType": {
                                "bool": {
                                  "nullability": "NULLABILITY_NULLABLE"
                                }
                              },
                              "arguments": [{
                                "value": {
                                  "selection": {
                                    "directReference": {
                                      "structField": {
                                        "field": 11
                                      }
                                    }
                                  }
                                }
                              }, {
                                "value": {
                                  "selection": {
                                    "directReference": {
                                      "structField": {
                                        "field": 12
                                      }
                                    }
                                  }
                                }
                              }]
                            }
                          }
                        }, {
                          "value": {
                            "scalarFunction": {
                              "functionReference": 1,
                              "outputType": {
                                "bool": {
                                  "nullability": "NULLABILITY_REQUIRED"
                                }
                              },
                              "arguments": [{
                                "value": {
                                  "selection": {
                                    "directReference": {
                                      "structField": {
                                        "field": 11
                                      }
                                    }
                                  }
                                }
                              }]
                            }
                          }
                        }]
                      }
                    }
                  }, {
                    "value": {
                      "scalarFunction": {
                        "functionReference": 1,
                        "outputType": {
                          "bool": {
                            "nullability": "NULLABILITY_REQUIRED"
                          }
                        },
                        "arguments": [{
                          "value": {
                            "selection": {
                              "directReference": {
                                "structField": {
                                  "field": 12
                                }
                              }
                            }
                          }
                        }]
                      }
                    }
                  }]
                }
              }
            }
          },
          "expressions": [{
            "scalarFunction": {
              "functionReference": 3,
              "outputType": {
                "i32": {
                  "nullability": "NULLABILITY_REQUIRED"
                }
              },
              "arguments": [{
                "value": {
                  "selection": {
                    "directReference": {
                      "structField": {
                      }
                    }
                  }
                }
              }]
            }
          }, {
            "selection": {
              "directReference": {
                "structField": {
                }
              }
            }
          }]
        }
      },
      "names": ["hash_partition_key#171", "l_orderkey#16"]
    }
  }]
}
*(3) Substrait plan: {
  "extensions": [{
    "extensionFunction": {
      "name": "equal:i64_i64"
    }
  }, {
    "extensionFunction": {
      "functionAnchor": 2,
      "name": "murmur3hash:opt_str"
    }
  }, {
    "extensionFunction": {
      "functionAnchor": 1,
      "name": "count:opt_i32"
    }
  }],
  "relations": [{
    "root": {
      "input": {
        "project": {
          "common": {
            "direct": {
            }
          },
          "input": {
            "aggregate": {
              "common": {
                "direct": {
                }
              },
              "input": {
                "project": {
                  "common": {
                    "direct": {
                    }
                  },
                  "input": {
                    "project": {
                      "common": {
                        "direct": {
                        }
                      },
                      "input": {
                        "join": {
                          "left": {
                            "read": {
                              "common": {
                                "direct": {
                                }
                              },
                              "baseSchema": {
                                "names": ["l_orderkey#16"],
                                "struct": {
                                  "types": [{
                                    "i64": {
                                      "nullability": "NULLABILITY_NULLABLE"
                                    }
                                  }]
                                }
                              },
                              "localFiles": {
                                "items": [{
                                  "uriFile": "iterator:0"
                                }]
                              }
                            }
                          },
                          "right": {
                            "read": {
                              "common": {
                                "direct": {
                                }
                              },
                              "baseSchema": {
                                "names": ["o_orderkey#56", "o_orderpriority#61"],
                                "struct": {
                                  "types": [{
                                    "i64": {
                                      "nullability": "NULLABILITY_NULLABLE"
                                    }
                                  }, {
                                    "string": {
                                      "nullability": "NULLABILITY_NULLABLE"
                                    }
                                  }]
                                }
                              },
                              "localFiles": {
                                "items": [{
                                  "uriFile": "iterator:1"
                                }]
                              }
                            }
                          },
                          "expression": {
                            "scalarFunction": {
                              "outputType": {
                                "bool": {
                                  "nullability": "NULLABILITY_NULLABLE"
                                }
                              },
                              "arguments": [{
                                "value": {
                                  "selection": {
                                    "directReference": {
                                      "structField": {
                                      }
                                    }
                                  }
                                }
                              }, {
                                "value": {
                                  "selection": {
                                    "directReference": {
                                      "structField": {
                                        "field": 1
                                      }
                                    }
                                  }
                                }
                              }]
                            }
                          },
                          "type": "JOIN_TYPE_RIGHT_SEMI",
                          "advancedExtension": {
                            "optimization": {
                              "@type": "/google.protobuf.StringValue",
                              "value": "JoinParameters:isBHJ\u003d0\nisNullAwareAntiJoin\u003d0\nbuildHashTableId\u003d\nisExistenceJoin\u003d0\n"
                            },
                            "enhancement": {
                              "@type": "/substrait.Type",
                              "struct": {
                                "types": [{
                                  "i64": {
                                    "nullability": "NULLABILITY_NULLABLE"
                                  }
                                }, {
                                  "i64": {
                                    "nullability": "NULLABILITY_NULLABLE"
                                  }
                                }, {
                                  "string": {
                                    "nullability": "NULLABILITY_NULLABLE"
                                  }
                                }],
                                "nullability": "NULLABILITY_REQUIRED"
                              }
                            }
                          }
                        }
                      },
                      "expressions": [{
                        "selection": {
                          "directReference": {
                            "structField": {
                            }
                          }
                        }
                      }, {
                        "selection": {
                          "directReference": {
                            "structField": {
                              "field": 1
                            }
                          }
                        }
                      }]
                    }
                  },
                  "expressions": [{
                    "selection": {
                      "directReference": {
                        "structField": {
                          "field": 1
                        }
                      }
                    }
                  }]
                }
              },
              "groupings": [{
                "groupingExpressions": [{
                  "selection": {
                    "directReference": {
                      "structField": {
                      }
                    }
                  }
                }]
              }],
              "measures": [{
                "measure": {
                  "functionReference": 1,
                  "phase": "AGGREGATION_PHASE_INITIAL_TO_INTERMEDIATE",
                  "outputType": {
                    "i64": {
                      "nullability": "NULLABILITY_REQUIRED"
                    }
                  },
                  "arguments": [{
                    "value": {
                      "literal": {
                        "i32": 1
                      }
                    }
                  }]
                }
              }]
            }
          },
          "expressions": [{
            "scalarFunction": {
              "functionReference": 2,
              "outputType": {
                "i32": {
                  "nullability": "NULLABILITY_REQUIRED"
                }
              },
              "arguments": [{
                "value": {
                  "selection": {
                    "directReference": {
                      "structField": {
                      }
                    }
                  }
                }
              }]
            }
          }, {
            "selection": {
              "directReference": {
                "structField": {
                }
              }
            }
          }, {
            "selection": {
              "directReference": {
                "structField": {
                  "field": 1
                }
              }
            }
          }]
        }
      },
      "names": ["hash_partition_key#172", "o_orderpriority#61", "count#169"]
    }
  }]
}
*(4) Substrait plan: {
  "extensions": [{
    "extensionFunction": {
      "name": "count:opt_i32"
    }
  }],
  "relations": [{
    "root": {
      "input": {
        "project": {
          "common": {
            "direct": {
            }
          },
          "input": {
            "aggregate": {
              "common": {
                "direct": {
                }
              },
              "input": {
                "read": {
                  "common": {
                    "direct": {
                    }
                  },
                  "baseSchema": {
                    "names": ["o_orderpriority#61", "count#169"],
                    "struct": {
                      "types": [{
                        "string": {
                          "nullability": "NULLABILITY_NULLABLE"
                        }
                      }, {
                        "i64": {
                          "nullability": "NULLABILITY_REQUIRED"
                        }
                      }]
                    }
                  },
                  "localFiles": {
                    "items": [{
                      "uriFile": "iterator:0"
                    }]
                  }
                }
              },
              "groupings": [{
                "groupingExpressions": [{
                  "selection": {
                    "directReference": {
                      "structField": {
                      }
                    }
                  }
                }]
              }],
              "measures": [{
                "measure": {
                  "phase": "AGGREGATION_PHASE_INTERMEDIATE_TO_RESULT",
                  "outputType": {
                    "i64": {
                      "nullability": "NULLABILITY_REQUIRED"
                    }
                  },
                  "arguments": [{
                    "value": {
                      "selection": {
                        "directReference": {
                          "structField": {
                            "field": 1
                          }
                        }
                      }
                    }
                  }]
                }
              }]
            }
          },
          "expressions": [{
            "selection": {
              "directReference": {
                "structField": {
                }
              }
            }
          }, {
            "selection": {
              "directReference": {
                "structField": {
                  "field": 1
                }
              }
            }
          }]
        }
      },
      "names": ["o_orderpriority#61", "order_count#122"]
    }
  }]
}
*(5) Substrait plan: {
  "relations": [{
    "root": {
      "input": {
        "sort": {
          "common": {
            "direct": {
            }
          },
          "input": {
            "read": {
              "common": {
                "direct": {
                }
              },
              "baseSchema": {
                "names": ["o_orderpriority#61", "order_count#122"],
                "struct": {
                  "types": [{
                    "string": {
                      "nullability": "NULLABILITY_NULLABLE"
                    }
                  }, {
                    "i64": {
                      "nullability": "NULLABILITY_REQUIRED"
                    }
                  }]
                }
              },
              "localFiles": {
                "items": [{
                  "uriFile": "iterator:0"
                }]
              }
            }
          },
          "sorts": [{
            "expr": {
              "selection": {
                "directReference": {
                  "structField": {
                  }
                }
              }
            },
            "direction": "SORT_DIRECTION_ASC_NULLS_FIRST"
          }]
        }
      },
      "names": ["o_orderpriority#61", "order_count#122"]
    }
  }]
}
+- == Initial Plan ==
   Sort [o_orderpriority#61 ASC NULLS FIRST], true, 0
   +- Exchange rangepartitioning(o_orderpriority#61 ASC NULLS FIRST, 384), ENSURE_REQUIREMENTS, [plan_id=69]
      +- HashAggregate(keys=[o_orderpriority#61], functions=[count(1)], output=[o_orderpriority#61, order_count#122L])
         +- Exchange hashpartitioning(o_orderpriority#61, 384), ENSURE_REQUIREMENTS, [plan_id=66]
            +- HashAggregate(keys=[o_orderpriority#61], functions=[partial_count(1)], output=[o_orderpriority#61, count#169L])
               +- Project [o_orderpriority#61]
                  +- ShuffledHashJoin [o_orderkey#56L], [l_orderkey#16L], LeftSemi, BuildRight
                     :- Exchange hashpartitioning(o_orderkey#56L, 384), ENSURE_REQUIREMENTS, [plan_id=60]
                     :  +- Project [o_orderkey#56L, o_orderpriority#61]
                     :     +- Filter ((isnotnull(o_orderdate#60) AND (o_orderdate#60 >= 1993-07-01)) AND (o_orderdate#60 < 1993-10-01))
                     :        +- BatchScan[o_orderkey#56L, o_orderdate#60, o_orderpriority#61] ParquetScan DataFilters: [isnotnull(o_orderdate#60), (o_orderdate#60 >= 1993-07-01), (o_orderdate#60 < 1993-10-01)], Format: parquet, Location: InMemoryFileIndex(1 paths)[hdfs://10.112.242.189:9000/tpch_sf2t_parquet_192p/orders], PartitionFilters: [], PushedFilters: [IsNotNull(o_orderdate), GreaterThanOrEqual(o_orderdate,1993-07-01), LessThan(o_orderdate,1993-10..., ReadSchema: struct<o_orderkey:bigint,o_orderdate:string,o_orderpriority:string>, PushedFilters: [IsNotNull(o_orderdate), GreaterThanOrEqual(o_orderdate,1993-07-01), LessThan(o_orderdate,1993-10-01)] RuntimeFilters: []
                     +- Exchange hashpartitioning(l_orderkey#16L, 384), ENSURE_REQUIREMENTS, [plan_id=61]
                        +- Project [l_orderkey#16L]
                           +- Filter (((l_commitdate#27 < l_receiptdate#28) AND isnotnull(l_commitdate#27)) AND isnotnull(l_receiptdate#28))
                              +- BatchScan[l_orderkey#16L, l_partkey#17L, l_suppkey#18L, l_linenumber#19, l_quantity#20, l_extendedprice#21, l_discount#22, l_tax#23, l_returnflag#24, l_linestatus#25, l_shipdate#26, l_commitdate#27, l_receiptdate#28, l_shipinstruct#29, l_shipmode#30, l_comment#31] ParquetScan DataFilters: [(l_commitdate#27 < l_receiptdate#28), isnotnull(l_commitdate#27), isnotnull(l_receiptdate#28)], Format: parquet, Location: InMemoryFileIndex(1 paths)[hdfs://10.112.242.189:9000/tpch_sf2t_parquet_192p/lineitem], PartitionFilters: [], PushedFilters: [IsNotNull(l_commitdate), IsNotNull(l_receiptdate)], ReadSchema: struct<l_orderkey:bigint,l_partkey:bigint,l_suppkey:bigint,l_linenumber:int,l_quantity:double,l_e..., PushedFilters: [IsNotNull(l_commitdate), IsNotNull(l_receiptdate)] RuntimeFilters: []
```
</details>

test by CH[[193]]

### CODE REVIEW

File: gluten-core/src/main/scala/io/glutenproject/execution/HashJoinExecTransformer.scala [hunk 1]
```diff
@@ -966,8 +964,12 @@ trait HashJoinLikeExecTransformer
       .append("buildHashTableId=").append(buildHashTableId).append("\n")
       .append("isExistenceJoin=").append(
       if (joinType.isInstanceOf[ExistenceJoin]) 1 else 0).append("\n")
+    val message = StringValue
+      .newBuilder()
+      .setValue(joinParametersStr.toString)
+      .build()
```
Comment (marin-ma) [inline gluten-core/src/main/scala/io/glutenproject/execution/HashJoinExecTransformer.scala:970]: @zzcclp In this PR we converted the substrait plan to json on java side, and printed the json string in df.explain(true). The to-json conversion on java side failed in our previous way, and is fixed by using StringValue to pack strings in Any. This also requires each backend to use the `UnpackTo` to parse this field. Could you help update in CH backend? Here's the update in Velox backend: https://github.com/oap-project/velox/pull/72/files#diff-6bbec4a6c456cb1efcf0c8f2dae7b80982e9002b9cf2a3460fd1ae545f291e49

**Review by rui-mo — APPROVED** [review]
LGTM.

File: gluten-core/src/main/scala/io/glutenproject/execution/HashJoinExecTransformer.scala (hunk 1 above)
Comment (zzcclp) [reply gluten-core/src/main/scala/io/glutenproject/execution/HashJoinExecTransformer.scala:970]: @liuneng1994 please help to review and update.

**Comment by marin-ma** [comment]
@liuneng1994 CH UT still failed. PTAL.

**Comment by FelixYBW** [comment]
With the PR, user can easily print out the substrait string then save into a json file. So one can easily reproduce a whole stage like Q4 or Q17, etc. from native benchmark. With the benchmark we can easily debug or profile the hotspot and do optimizations

**Comment by FelixYBW** [comment]
@marin-ma  can you add a readme under docs to show the steps hot to generate the query plan, generate data, then run benchmark to reproduce the stage? It's even better if we can create a notebook to do this. 

The scala code is not familar to some developers if they just want to optimize c++ code.

**Comment by marin-ma** [comment]
> @marin-ma can you add a readme under docs to show the steps hot to generate the query plan, generate data, then run benchmark to reproduce the stage? It's even better if we can create a notebook to do this.
> 
> The scala code is not familar to some developers if they just want to optimize c++ code.

Add doc https://github.com/oap-project/gluten/pull/579

**Pull request merged**